File: tests/sdk/go/storage_test.go
package sdk_test

import (
	"encoding/json"
	"io"
	"net/http"
	"net/url"
	"strings"
	"testing"
	"time"
)

func TestListBucketsSmoke(t *testing.T) {
	ep := requireReady(t)
	token := requireToken(t)
	project := projectID()

	status, body := doJSON(t, http.MethodGet, ep+"/storage/v1/b?project="+url.QueryEscape(project), token, nil)
	if status != http.StatusOK {
		t.Fatalf("list buckets status=%d body=%s", status, body)
	}
	var parsed map[string]any
	if err := json.Unmarshal(body, &parsed); err != nil {
		t.Fatalf("decode buckets: %v body=%s", err, body)
	}
	if kind, _ := parsed["kind"].(string); kind != "storage#buckets" {
		t.Fatalf("kind=%v want storage#buckets body=%s", parsed["kind"], body)
	}
}

func TestGCSGenerateSignedURLSmoke(t *testing.T) {
	ep := requireReady(t)
	token := requireToken(t)
	project := projectID()
	bucket := uniqueID("sdk-signed")
	bucketPath := ep + "/storage/v1/b/" + url.PathEscape(bucket)

	status, body := doJSON(t, http.MethodPost, ep+"/storage/v1/b?project="+url.QueryEscape(project), token, map[string]any{
		"name": bucket,
	})
	if status != http.StatusOK {
		t.Fatalf("create bucket status=%d body=%s", status, body)
	}
	t.Cleanup(func() {
		_, _, _ = doJSONErr(http.MethodDelete, bucketPath, token, nil)
	})

	status, body = doJSON(t, http.MethodPost, bucketPath+"/o/smoke.txt:generateSignedUrl", token, map[string]any{
		"method":  "GET",
		"expires": 600,
		"alt":     "media",
	})
	if status != http.StatusOK {
		t.Fatalf("generateSignedUrl status=%d body=%s", status, body)
	}
	var parsed map[string]any
	if err := json.Unmarshal(body, &parsed); err != nil {
		t.Fatalf("decode generateSignedUrl: %v body=%s", err, body)
	}
	if got, _ := parsed["signedUrl"].(string); got == "" {
		t.Fatalf("missing signedUrl body=%s", body)
	}
}

func TestGCSRetentionDeleteDenySmoke(t *testing.T) {
	ep := requireReady(t)
	token := requireToken(t)
	project := projectID()
	bucket := uniqueID("sdk-retain")
	bucketPath := ep + "/storage/v1/b/" + url.PathEscape(bucket)
	objectPath := bucketPath + "/o/" + url.PathEscape("held.txt")

	status, body := doJSON(t, http.MethodPost, ep+"/storage/v1/b?project="+url.QueryEscape(project), token, map[string]any{
		"name": bucket,
	})
	if status != http.StatusOK {
		t.Fatalf("create bucket status=%d body=%s", status, body)
	}
	t.Cleanup(func() {
		_, _, _ = doJSONErr(http.MethodDelete, objectPath, token, nil)
		_, _, _ = doJSONErr(http.MethodDelete, bucketPath, token, nil)
	})

	status, body = doJSON(t, http.MethodPatch, bucketPath, token, map[string]any{
		"retentionPolicy": map[string]any{"retentionPeriod": "3600"},
	})
	if status != http.StatusOK {
		t.Fatalf("patch retention status=%d body=%s", status, body)
	}

	upReq, err := http.NewRequest(http.MethodPost, ep+"/upload/storage/v1/b/"+url.PathEscape(bucket)+"/o?uploadType=media&name="+url.QueryEscape("held.txt"), strings.NewReader("held"))
	if err != nil {
		t.Fatalf("upload request: %v", err)
	}
	upReq.Header.Set("Authorization", "Bearer "+token)
	upReq.Header.Set("Content-Type", "text/plain")
	upResp, err := (&http.Client{Timeout: 5 * time.Second}).Do(upReq)
	if err != nil {
		t.Fatalf("upload: %v", err)
	}
	upBody, _ := io.ReadAll(upResp.Body)
	_ = upResp.Body.Close()
	if upResp.StatusCode != http.StatusOK {
		t.Fatalf("upload status=%d body=%s", upResp.StatusCode, upBody)
	}

	status, body = doJSON(t, http.MethodDelete, objectPath, token, nil)
	if status == http.StatusOK {
		t.Fatalf("delete under retention should fail; got status=%d body=%s", status, body)
	}
	var errBody map[string]any
	if err := json.Unmarshal(body, &errBody); err != nil {
		t.Fatalf("decode delete error: %v body=%s", err, body)
	}
	e, _ := errBody["error"].(map[string]any)
	if e == nil || e["status"] != "FAILED_PRECONDITION" {
		t.Fatalf("want FAILED_PRECONDITION, got status=%d body=%s", status, body)
	}
}

func TestListBigtableInstancesSmoke(t *testing.T) {
	ep := requireReady(t)
	token := requireToken(t)
	project := projectID()

	path := ep + "/v2/projects/" + project + "/instances"
	status, body := doJSON(t, http.MethodGet, path, token, nil)
	if status != http.StatusOK {
		t.Fatalf("list bigtable instances status=%d body=%s", status, body)
	}
	var parsed map[string]any
	if err := json.Unmarshal(body, &parsed); err != nil {
		t.Fatalf("decode bigtable instances: %v body=%s", err, body)
	}
	if _, ok := parsed["instances"]; !ok {
		t.Fatalf("missing instances field body=%s", body)
	}
}

func TestListMemorystoreInstancesSmoke(t *testing.T) {
	ep := requireReady(t)
	token := requireToken(t)
	project := projectID()

	path := ep + "/v1/projects/" + project + "/locations/us-central1/instances"
	status, body := doJSON(t, http.MethodGet, path, token, nil)
	if status != http.StatusOK {
		t.Fatalf("list memorystore instances status=%d body=%s", status, body)
	}
	var parsed map[string]any
	if err := json.Unmarshal(body, &parsed); err != nil {
		t.Fatalf("decode memorystore instances: %v body=%s", err, body)
	}
	if _, ok := parsed["instances"]; !ok {
		t.Fatalf("missing instances field body=%s", body)
	}
}

func TestListManagedKafkaClustersSmoke(t *testing.T) {
	ep := requireReady(t)
	token := requireToken(t)
	project := projectID()

	path := ep + "/v1/projects/" + project + "/locations/us-central1/clusters"
	status, body := doJSON(t, http.MethodGet, path, token, nil)
	if status != http.StatusOK {
		t.Fatalf("list managed kafka clusters status=%d body=%s", status, body)
	}
	var parsed map[string]any
	if err := json.Unmarshal(body, &parsed); err != nil {
		t.Fatalf("decode managed kafka clusters: %v body=%s", err, body)
	}
	if _, ok := parsed["clusters"]; !ok {
		t.Fatalf("missing clusters field body=%s", body)
	}
}

func TestListCloudSQLInstancesSmoke(t *testing.T) {
	ep := requireReady(t)
	token := requireToken(t)
	project := projectID()

	path := ep + "/sql/v1/projects/" + project + "/instances"
	status, body := doJSON(t, http.MethodGet, path, token, nil)
	if status != http.StatusOK {
		t.Fatalf("list cloudsql instances status=%d body=%s", status, body)
	}
	var parsed map[string]any
	if err := json.Unmarshal(body, &parsed); err != nil {
		t.Fatalf("decode cloudsql instances: %v body=%s", err, body)
	}
	if _, ok := parsed["items"]; !ok {
		t.Fatalf("missing items field body=%s", body)
	}
}

func TestListBigQueryDatasetsSmoke(t *testing.T) {
	ep := requireReady(t)
	token := requireToken(t)
	project := projectID()

	path := ep + "/bigquery/v2/projects/" + project + "/datasets"
	status, body := doJSON(t, http.MethodGet, path, token, nil)
	if status != http.StatusOK {
		t.Fatalf("list BigQuery datasets status=%d body=%s", status, body)
	}
	var parsed map[string]any
	if err := json.Unmarshal(body, &parsed); err != nil {
		t.Fatalf("decode BigQuery datasets: %v body=%s", err, body)
	}
	if _, ok := parsed["datasets"]; !ok {
		t.Fatalf("missing datasets field body=%s", body)
	}
}

func TestListSpannerInstancesSmoke(t *testing.T) {
	ep := requireReady(t)
	token := requireToken(t)
	project := projectID()

	path := ep + "/v1/projects/" + project + "/instances"
	status, body := doJSON(t, http.MethodGet, path, token, nil)
	if status != http.StatusOK {
		t.Fatalf("list Spanner instances status=%d body=%s", status, body)
	}
	var parsed map[string]any
	if err := json.Unmarshal(body, &parsed); err != nil {
		t.Fatalf("decode Spanner instances: %v body=%s", err, body)
	}
	if _, ok := parsed["instances"]; !ok {
		t.Fatalf("missing instances field body=%s", body)
	}
}

File: tests/sdk/nodejs/storage.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import { doJSON, hasOwn, projectID, requireReady, requireToken, uniqueID } from "./helpers.mjs";

test("list buckets smoke", async (t) => {
  const ep = await requireReady(t);
  if (!ep) return;
  const token = requireToken(t);
  if (!token) return;
  const project = projectID();

  const { status, body } = await doJSON(
    "GET",
    `${ep}/storage/v1/b?project=${encodeURIComponent(project)}`,
    token,
  );
  assert.equal(status, 200, `list buckets status=${status} body=${body}`);
  const parsed = JSON.parse(body);
  assert.equal(parsed.kind, "storage#buckets", body);
});

test("gcs generateSignedUrl smoke", async (t) => {
  const ep = await requireReady(t);
  if (!ep) return;
  const token = requireToken(t);
  if (!token) return;
  const project = projectID();
  const bucket = uniqueID("sdk-signed");
  const bucketPath = `${ep}/storage/v1/b/${encodeURIComponent(bucket)}`;

  const created = await doJSON(
    "POST",
    `${ep}/storage/v1/b?project=${encodeURIComponent(project)}`,
    token,
    { name: bucket },
  );
  assert.equal(created.status, 200, `create bucket status=${created.status} body=${created.body}`);
  t.after(async () => {
    try {
      await doJSON("DELETE", bucketPath, token);
    } catch {
      /* best-effort cleanup */
    }
  });

  const { status, body } = await doJSON(
    "POST",
    `${bucketPath}/o/smoke.txt:generateSignedUrl`,
    token,
    { method: "GET", expires: 600, alt: "media" },
  );
  assert.equal(status, 200, `generateSignedUrl status=${status} body=${body}`);
  const parsed = JSON.parse(body);
  assert.ok(parsed.signedUrl, body);
});

test("gcs retention delete deny smoke", async (t) => {
  const ep = await requireReady(t);
  if (!ep) return;
  const token = requireToken(t);
  if (!token) return;
  const project = projectID();
  const bucket = uniqueID("sdk-retain");
  const bucketPath = `${ep}/storage/v1/b/${encodeURIComponent(bucket)}`;
  const objectPath = `${bucketPath}/o/${encodeURIComponent("held.txt")}`;

  const created = await doJSON(
    "POST",
    `${ep}/storage/v1/b?project=${encodeURIComponent(project)}`,
    token,
    { name: bucket },
  );
  assert.equal(created.status, 200, `create bucket status=${created.status} body=${created.body}`);
  t.after(async () => {
    try {
      await doJSON("DELETE", objectPath, token);
    } catch {
      /* best-effort cleanup */
    }
    try {
      await doJSON("DELETE", bucketPath, token);
    } catch {
      /* best-effort cleanup */
    }
  });

  const patched = await doJSON("PATCH", bucketPath, token, {
    retentionPolicy: { retentionPeriod: "3600" },
  });
  assert.equal(patched.status, 200, `patch retention status=${patched.status} body=${patched.body}`);

  const upRes = await fetch(
    `${ep}/upload/storage/v1/b/${encodeURIComponent(bucket)}/o?uploadType=media&name=${encodeURIComponent("held.txt")}`,
    {
      method: "POST",
      headers: {
        Authorization: `Bearer ${token}`,
        "Content-Type": "text/plain",
      },
      body: "held",
      signal: AbortSignal.timeout(5000),
    },
  );
  const upBody = await upRes.text();
  assert.equal(upRes.status, 200, `upload status=${upRes.status} body=${upBody}`);

  const deleted = await doJSON("DELETE", objectPath, token);
  assert.notEqual(deleted.status, 200, `delete under retention should fail; body=${deleted.body}`);
  const errBody = JSON.parse(deleted.body);
  assert.equal(errBody.error?.status, "FAILED_PRECONDITION", deleted.body);
});

test("list bigtable instances smoke", async (t) => {
  const ep = await requireReady(t);
  if (!ep) return;
  const token = requireToken(t);
  if (!token) return;
  const project = projectID();

  const { status, body } = await doJSON("GET", `${ep}/v2/projects/${project}/instances`, token);
  assert.equal(status, 200, `list bigtable instances status=${status} body=${body}`);
  const parsed = JSON.parse(body);
  assert.ok(hasOwn(parsed, "instances"), body);
});

test("list memorystore instances smoke", async (t) => {
  const ep = await requireReady(t);
  if (!ep) return;
  const token = requireToken(t);
  if (!token) return;
  const project = projectID();

  const { status, body } = await doJSON(
    "GET",
    `${ep}/v1/projects/${project}/locations/us-central1/instances`,
    token,
  );
  assert.equal(status, 200, `list memorystore instances status=${status} body=${body}`);
  const parsed = JSON.parse(body);
  assert.ok(hasOwn(parsed, "instances"), body);
});

test("list cloud sql instances smoke", async (t) => {
  const ep = await requireReady(t);
  if (!ep) return;
  const token = requireToken(t);
  if (!token) return;
  const project = projectID();

  const { status, body } = await doJSON(
    "GET",
    `${ep}/sql/v1/projects/${project}/instances`,
    token,
  );
  assert.equal(status, 200, `list cloudsql instances status=${status} body=${body}`);
  const parsed = JSON.parse(body);
  assert.ok(hasOwn(parsed, "items"), body);
});

test("list managed kafka clusters smoke", async (t) => {
  const ep = await requireReady(t);
  if (!ep) return;
  const token = requireToken(t);
  if (!token) return;
  const project = projectID();

  const { status, body } = await doJSON(
    "GET",
    `${ep}/v1/projects/${project}/locations/us-central1/clusters`,
    token,
  );
  assert.equal(status, 200, `list managed kafka clusters status=${status} body=${body}`);
  const parsed = JSON.parse(body);
  assert.ok(hasOwn(parsed, "clusters"), body);
});

test("list bigquery datasets smoke", async (t) => {
  const ep = await requireReady(t);
  if (!ep) return;
  const token = requireToken(t);
  if (!token) return;
  const project = projectID();

  const { status, body } = await doJSON(
    "GET",
    `${ep}/bigquery/v2/projects/${project}/datasets`,
    token,
  );
  assert.equal(status, 200, `list BigQuery datasets status=${status} body=${body}`);
  const parsed = JSON.parse(body);
  assert.ok(hasOwn(parsed, "datasets"), body);
});

test("list spanner instances smoke", async (t) => {
  const ep = await requireReady(t);
  if (!ep) return;
  const token = requireToken(t);
  if (!token) return;
  const project = projectID();

  const { status, body } = await doJSON("GET", `${ep}/v1/projects/${project}/instances`, token);
  assert.equal(status, 200, `list Spanner instances status=${status} body=${body}`);
  const parsed = JSON.parse(body);
  assert.ok(hasOwn(parsed, "instances"), body);
});
